### PR TITLE
📝 Feature/tuerme

### LINKED ISSUE #17: Tower example
EDIT: Renamed from hanoi to tower. Real Hanoi has the extra condition that the pieces have different weights and pieces with more weight can only be at the bottom. That might make for an interesting example in the future (interesting path finding algorithm, might also work nicely with processengine if modelled correctly).

Came up with an idea for the tower example.
Objective is to move the tower from 1 to 3.
You can technically only move one piece at a time.
Solution is to flip the tower onto 2, then flip again onto 3.
```
5 
m
i
n
d
s
- - -
1 2 3
This is how it starts.

m
i
n
d
s 5
- - -
1 2 3
The first move.

  s
  d
  n
  i
  m
  5
- - -
1 2 3
Multiple moves make up the flip.

    5 
    m
    i
    n
    d
    s
- - -
1 2 3
And another flip.
```
## Why I think the idea is good

It's a riddle, riddles are fun and we can show how to make clean solutions for a task which might seem complicated at first.

While algorithms are usually something for the code side, this has at least some relation to real world situations(storagemanagment & planning).

## BPMN-Design

2 diagrams for this:

### **Move tower**

Needs mostly two call activities(for the two flips which make up the moving of a tower).
Would add some more tasks(usertask for input which tower to move where, maybe errorhandling for user task).
### **Flip Tower**

Also very simple. token contains originalPosition and targetPosition
1. Move one block from originalPosition to targetPosition (http-request)
2. Check if originalPosition is free (http-request)
   - if yes, quit task
   - if no go to 1.

Would do this with just 2 tasks + 1 xor.

##  backend

Needs a backend in order to display the current situation and keep state.
2 routes required.
```
GET /is-free/:position
POST /move/:from/:to
```
Would render the towers in terminal in first step. We can also add a simple web page.

### PR DESCRIPTION
**Pull request by LeoTT**
Add towers example
---

**Changes:**

1. Adds diagrams
2. Adds readme

**Issues:**

Closes #17 
PR: #PullRequest

---

See the [examples](#example) for inspiration.

## How can others test the changes?

Read readme, test diagrams (run `Move Tower`-Diagramm)

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.).
- [x] I've rebased the `develop` branch with my branch before finishing this PR.
- [x] I've **summarized all changes** in a list above.
- [x] I've mentioned all **PRs, which relate to this one**.
- [x] I've prefixed my Pull Request title is according to [gitmoji guide](https://gitmoji.carloscuesta.me/).

## Example

<details>
<summary>
:warning: Before merging please click here to expand and follow the guide.
</summary>
<br>

Please use `:twisted_rightwards_arrows:` at the beginning of your merge commit title.

Example 1:

<pre>
<code>
:twisted_rightwards_arrows: :bug: Fix Wrong Text Decoration at ...
</code>
</pre>

To get your commit message, just copy the first part of this pull request.

Example 2:
<pre>
<code>
**Changes:**

- Fixes Wrong Text Decoration at ...
- Fixes some typos
- ...

**Issues:**

Closes #NumberOfFixedIssue

PR: #NumberOfThisPR
</code>
</pre>
</details>

## Emojis

<details>
<summary>
Expand for a list of most used Emojis.
</summary>
<br>

Please prefix your commit messages with an Emoji.

Ref: https://gitmoji.carloscuesta.me/

| Description              | Glyphe               | Emoji  |
|--------------------------|----------------------|--------|
| Bugfix                   | `:bug:`              | 🐛     |
| Configuration releated   | `:wrench:`           | 🔧     |
| Cosmetic                 | `:lipstick:`         | 💄     |
| Dependencies Downgrade   | `:arrow_down:`       | ⬇️     |
| Dependencies Upgrade     | `:arrow_up:`         | ⬆️     |
| Formatting               | `:art:`              | 🎨     |
| Improving Performance    | `:zap:`              | ⚡️     |
| Initial commit           | `:tada:`             | 🎉     |
| Linter                   | `:rotating_light:`   | 🚨     |
| Miscellaneous            | `:package:`          | 📦     |
| New Feature              | `:sparkles:`         | ✨     |
| Refactoring Code         | `:recycle:`          | ♻️     |
| Releasing / Version tags | `:bookmark:`         | 🔖     |
| Removing Stuff           | `:fire:`             | 🔥     |
| Tests                    | `:white_check_mark:` | ✅     |
| Work In Progress (WIP)   | `:construction:`     | 🚧     |

</details>

### CODE REVIEW

File: solutions/Towers/README.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+Prozesses. Was wäre, wenn folgend die Aufgabe entsteht, dass der Turm
+zur Position `1` verschoben werden soll?  Und wie sähe das Diagramm
+aus, wenn der Turm nicht nur drei Elemente hoch wäre, sondern
+hunderte?
+
+## Lösungsidee
+
+Zur modularen Lösung erstellen wir drei Prozesse:
+
+1. **Move Element** erlaubt das Bewegen des obersten Element eines
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: typo: des obersten Element**s**

File: solutions/Towers/README.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+Um komfortabel auf diese Startparameter zugreifen zu können, vergeben
+wir eine neue ID `startevent_arguments`.
+
+
+## ScriptTask `Take element of one tower`
+
+Die eigentliche Funktionalität bewältigen wir mit ScriptTasks.
+
+Wir hängen einen neuen ScriptTask an das StartEvent mit dem Name
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: typo: mit dem Name**n**

File: solutions/Towers/README.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+  };
+})();
+```
+
+Der Programmcode muss in das Feld `Script` im PropertyPanel eingefügt
+werden.
+
+## ScriptTask `Put element onto another tower`
+
+Wir erstellen einen weiteren ScriptTask mit dem Name `Put element
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: typo: mit dem Name**n**

File: solutions/Towers/README.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+  const {tower, element} = token.history.scripttask_take_element;
+  const {toIndex} = token.history.startevent_arguments;
+  tower[toIndex].push(element);
+  return tower;
+})();
+```
+
+Der ScriptTask wird mit dem EndEvent verknüpft. Was hier als letzter
+Tokenwert zurückgegeben wird, wird auch der resultierende Wert beim
+Aufruf der CallActivity sein.
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: Besser:
`Der Wert der am EndEvent ankommt wird auch als Ergebnis der CallActivity zurück gegeben.`

File: solutions/Towers/README.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+  tower: token.history.startevent_arguments.tower,
+  fromIndex: token.history.startevent_arguments.fromIndex,
+  toIndex: token.history.startevent_arguments.toIndex,
+}
+```
+
+> Zur Erinnerung: Beim **Move Element**-Diagramm haben wir dieses Format zur
+> Übersicht per Text-Annotation vermerkt.
+
+Wir vergeben den Name `Prepare parameters`.
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: typo: den Name**n**

File: solutions/Towers/README.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+
+## CallActivity `Move Element`
+
+Anschließend erstellen wir die CallActivity. Wir verlinken das zuvor
+erstellte `Move-Element`-Diagramm und setzen den Name auf `Move
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: den Name**n**

File: solutions/Towers/README.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+Der Prozess soll dann stoppen, wenn der Ausgangsturm an der Position
+`fromIndex` leer ist. Dies können wir mit folgendem Programmcode
+überprüfen:
+
+```js
+return token.history.startevent_arguments.tower[token.history.startevent_arguments.fromIndex].length === 0;
+```
+
+Wir versehen den ScriptTask mit der ID `servicetask_check_if_done` und
+dem Name `Check if original tower is empty`.
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: dem Name**n**

File: solutions/Towers/README.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+
+## ExclusiveGateway
+
+Dem ScriptTask folgend erstellen wir ein Gateway mit dem Name `Is
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: dem Name**n**

File: solutions/Towers/README.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+CallActivity angezeigt wird.
+
+Wir erstellen einen ScriptTask, wobei wir im PropertyPanel bei
+`Script` den folgenden Code eingeben:
+
+```js
+return token.history.callactivity_move_element;
+```
+
+Der ScriptTask erhält den Name `Return towers`.
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: den Name**n**

File: solutions/Towers/README.md [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+
+
+Wir fügen zudem einen Flow vom Gateway zum ScriptTask mit der Condition
+`token.history.servicetask_check_if_done === true` und dem Name `yes`
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: dem Name**n**

File: solutions/Towers/README.md [hunk 11]
```diff
@@ -0,0 +1,392 @@
...
+
+Der ScriptTask wird mit einem EndEvent verbunden.
+
+### Schleife
+
+Bis die Abbruchsbedingung erfüllt ist wollen wir weitere Elemente
+bewegen.
+
+Ausgehend vom Gateway erstellen wir einen Flow, welcher zurück zum
+ScriptTask mit dem Name `Prepare parameters` führt.  Der Flow wird mit
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: dem Name**n**

File: solutions/Towers/README.md [hunk 12]
```diff
@@ -0,0 +1,392 @@
...
+
+## ScriptTask `Create towers`
+
+Wir erstellen die Türme mit einem ScriptTask mit dem Code:
+
+```js
+return [["A", "B", "C"], [], []];
+```
+
+Der ScriptTask erhält von uns den Name `Create towers` und die ID
```
Comment (ElRaptorus) [inline solutions/Towers/README.md]: den Name**n**

File: solutions/Towers/README.md (hunk 2 above)
Comment (LeoTT) [reply solutions/Towers/README.md]: Wollte die englische Bezeichnung nehmen um klarer zu machen, dass es eine Property des Elements ist.
Liest sich aber echt nur wie ein typo, werds überall fixen.

**Pull request merged**